pmlc/dialect/pxa/transforms/stencil_generic.h

```diff
@@ -0,0 +1,114 @@
+// Copyright 2020 Intel Corporation
+
+#pragma once
+
+// TODO: includes etc
+#include <algorithm>
+#include <functional>
+#include <limits>
+#include <list>
+#include <map>
+#include <utility>
+#include <vector>
+
+// TODO: Just seeing if the stencil.cc includes work
+#include "mlir/ADT/TypeSwitch.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassOptions.h"
+
+#include "pmlc/dialect/eltwise/ir/ops.h"
+#include "pmlc/dialect/pxa/analysis/strides.h"
+#include "pmlc/dialect/pxa/ir/ops.h"
+#include "pmlc/dialect/pxa/transforms/passes.h"
+#include "pmlc/dialect/tile/ir/ops.h"
+
+#include "pmlc/util/logging.h"
+#include "pmlc/util/util.h"
+
+namespace pmlc::dialect::pxa {
+
+using BlockArgumentSet = llvm::SmallPtrSet<mlir::BlockArgument, 8>;
+
+struct TensorAndIndexPermutation {
+  // An order of the Tensors and Indexes used in an operation
+  llvm::SmallVector<Value, 3> tensors;
+  llvm::SmallVector<mlir::BlockArgument, 8> indexes;
+
+  TensorAndIndexPermutation() = default;
+
+  TensorAndIndexPermutation(llvm::SmallVector<Value, 3> tensors,
+                            llvm::SmallVector<mlir::BlockArgument, 8> indexes)
+      : tensors(tensors), indexes(indexes) {}
+};
+
+struct LoadStoreOps {
+  // The load and store ops of an AffineParallel
+  llvm::SmallVector<mlir::AffineLoadOp, 2> loads;
+  llvm::SmallVector<AffineReduceOp, 1>
+      stores; // TODO: Might be either store or reduce
+};
+
+// TODO: May want something more like an autotile.h UnionGenerator?
+using TileSizeGenerator = std::function<std::vector<size_t>(size_t)>;
```

Comment:
UnionGenerator is a TileSizeGenerator already! Specifically, you can pass a functor object to an std::function... See the following example code (which prints 15)

#include
#include

struct Foo {
double add;
double operator()(double val) { return val + add; }
};

int main() {
Foo x = {5};
std::function<double(double)> f = x;
std::cout << f(10) << "\n";
}

Reply:
Oh cool, thanks! (technically there's a `size_t`/`int64_t` problem, but that's not what my concern was and also something I need to be generally resolving with this code.)